File: ADMIN/Example-API-Flutter/Example.UserService.WebApp/Helpers/ExamPeriodTimeHelper.cs
using Example.Common.Enums;
using Example.Common.Utilities;
using Example.UserService.API.Models;

namespace Example.UserService.WebApp.Helpers
{
    /// <summary>Trạng thái thời gian kỳ thi trên UI WebApp (chưa đến giờ / đang mở / đã kết thúc).</summary>
    public enum ExamPeriodAvailability
    {
        NotYetStarted,
        Open,
        Ended
    }

    public static class ExamPeriodTimeHelper
    {
        public static ExamPeriodAvailability GetAvailability(ExamPeriodAssignmentModel assignment)
        {
            // API đã đóng kỳ → coi như kết thúc dù chưa hết giờ local
            if (assignment.ExamPeriodStatus == (int)ExamPeriodStatusEnum.Closed)
                return ExamPeriodAvailability.Ended;

            var now = VietnamTimeHelper.ToVietnamTime(DateTimeOffset.UtcNow);

            if (assignment.ExamPeriodEndAt.HasValue)
            {
                var end = VietnamTimeHelper.ToVietnamTime(assignment.ExamPeriodEndAt.Value);
                if (now > end)
                    return ExamPeriodAvailability.Ended;
            }

            if (assignment.ExamPeriodStartAt.HasValue)
            {
                var start = VietnamTimeHelper.ToVietnamTime(assignment.ExamPeriodStartAt.Value);
                if (now < start)
                    return ExamPeriodAvailability.NotYetStarted;
            }

            return ExamPeriodAvailability.Open;
        }

        public static bool CanStartExam(ExamPeriodAssignmentModel assignment)
        {
            if (assignment.Status == (int)ExamPeriodAssignmentStatusEnum.Completed
                || assignment.Status == (int)ExamPeriodAssignmentStatusEnum.Cancelled)
                return false;

            // Chỉ cho bấm "Bắt đầu" khi đang trong khung giờ kỳ thi
            return GetAvailability(assignment) == ExamPeriodAvailability.Open;
        }

        public static bool CanContinueExam(ExamPeriodAssignmentModel assignment)
        {
            if (assignment.Status == (int)ExamPeriodAssignmentStatusEnum.Completed
                || assignment.Status == (int)ExamPeriodAssignmentStatusEnum.Cancelled)
                return false;

            // Cho tiếp tục khi chưa hết kỳ (kể cả trước giờ mở — nếu đã có session InProgress từ trước)
            return GetAvailability(assignment) != ExamPeriodAvailability.Ended;
        }
    }
}

File: lib/core/utils/exam_period_time_helper.dart
import '../enums/app_enums.dart';
import '../../models/exam/exam_period_assignment.dart';
import '../../l10n/app_strings.dart';
import 'vietnam_time_helper.dart';

/// Helper kiểm tra khung giờ kỳ thi — port từ `ExamPeriodTimeHelper.cs` (WebApp).
class ExamPeriodTimeHelper {
  static ExamPeriodAvailability getAvailability(ExamPeriodAssignment assignment) {
    if (assignment.examPeriodStatus == 3) {
      return ExamPeriodAvailability.ended;
    }

    final now = VietnamTimeHelper.nowVietnam();

    if (assignment.examPeriodEndAt != null) {
      final end = VietnamTimeHelper.toVietnamWallClock(assignment.examPeriodEndAt!);
      if (now.isAfter(end)) return ExamPeriodAvailability.ended;
    }

    if (assignment.examPeriodStartAt != null) {
      final start =
          VietnamTimeHelper.toVietnamWallClock(assignment.examPeriodStartAt!);
      if (now.isBefore(start)) return ExamPeriodAvailability.notYetStarted;
    }

    return ExamPeriodAvailability.open;
  }

  static bool canStartExam(ExamPeriodAssignment assignment) {
    if (assignment.status == 3 || assignment.status == 5) return false;
    return getAvailability(assignment) == ExamPeriodAvailability.open;
  }

  static bool canContinueExam(ExamPeriodAssignment assignment) {
    if (assignment.status == 3 || assignment.status == 5) return false;
    return getAvailability(assignment) != ExamPeriodAvailability.ended;
  }

  static String examPeriodStatusName(ExamPeriodAssignment assignment) {
    final s = S.current;
    return switch (getAvailability(assignment)) {
      ExamPeriodAvailability.notYetStarted => s.examPeriodPublished,
      ExamPeriodAvailability.open => s.examPeriodPublished,
      ExamPeriodAvailability.ended => s.examPeriodClosed,
    };
  }

  static String displayStatusName(ExamPeriodAssignment assignment) {
    return S.current.displayStatusName(
      assignment.status,
      getAvailability(assignment),
    );
  }
}
